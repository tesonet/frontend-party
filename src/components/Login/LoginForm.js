import styled from 'styled-components';

export default styled.form`
  display: flex;
  flex-direction: column;
  margin: 0 auto;
  position: relative;
  @media (max-width: 400px) {
    width: 90%;
  }
`;
