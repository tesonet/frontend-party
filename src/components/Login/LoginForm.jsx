import React, {
    Fragment, useRef, memo,
} from 'react';
import PropTypes from 'prop-types';
import { useSelector } from 'react-redux';
import { Field, reduxForm } from 'redux-form';
import { Overlay, Tooltip } from 'react-bootstrap';
import LoginSubmitButton from './LoginSubmitButton';
import PasswordFieldWrapper from './Fields/PasswordFieldWrapper';
import UsernameFieldWrapper from './Fields/UsernameFieldWrapper';
import validate from '../../utils/validate';

const Form = (props) => {
    const { handleSubmit, error } = props;
    const submitted = useSelector((state) => state.auth.submitted);
    const formRef = useRef(null);

    return (
        <Fragment>
            <form
                onSubmit={handleSubmit}
                ref={formRef}
            >
                <Field
                    name="username"
                    component={UsernameFieldWrapper}
                    label="Username"
                />
                <Field
                    name="password"
                    component={PasswordFieldWrapper}
                    label="Password"
                />
                <LoginSubmitButton
                    disabled={!submitted}
                    className="login-button"
                />
            </form>
            <Overlay target={formRef.current} show={!!error} placement="bottom">
                {(tooltipProps) => (
                    <Tooltip id="button-tooltip" {...tooltipProps}>
                        {error}
                    </Tooltip>
                )}
            </Overlay>
        </Fragment>
    );
};

Form.defaultProps = {
    error: null,
};

Form.propTypes = {
    error: PropTypes.string,
    handleSubmit: PropTypes.func.isRequired,
};

const LoginForm = reduxForm({
    form: 'login',
    validate,
})(Form);

export default memo(LoginForm);
