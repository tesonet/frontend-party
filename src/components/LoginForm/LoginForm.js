import React, { useState } from "react";
import { useSelector, useDispatch } from "react-redux";

//Components.
import ImageProxy from "../ImageProxy/ImageProxy";

//Constants.
import { API_URL } from "../../constants/api";

//Images.
import testioImgPath from "../../assets/icons/logo.svg";

const LoginForm = () => {
  const [username, setUsername] = useState("");
  const [psw, setPsw] = useState("");
  const dispatch = useDispatch();

  const handleUsernameChange = e => {
    setUsername(e.target.value);
  };

  const handlePswChange = e => {
    setPsw(e.target.value);
  };

  const handleSubmit = e => {
    e.preventDefault();
    fetch(API_URL, {
      headers: { "content-type": "application/json; charset=UTF-8" },
      body: JSON.stringify({
        username: username,
        password: psw
      }),
      method: "POST"
    })
      .then(response => response.json())
      .then(data => {
        dispatch({ type: "login", payload: { data } });
      });
  };

  return (
    <React.Fragment>
      <ImageProxy src={testioImgPath} alt="testio" />
      <form className="LoginForm" onSubmit={handleSubmit}>
        <div className="LoginForm__input">
          <input
            type="text"
            placeholder="Username"
            value={username}
            onChange={handleUsernameChange}
          />
        </div>
        <div className="LoginForm__input">
          <input
            type="password"
            placeholder="Password"
            value={psw}
            onChange={handlePswChange}
          />
        </div>
        <input type="submit" value="Log In" />
      </form>
    </React.Fragment>
  );
};

export default LoginForm;
