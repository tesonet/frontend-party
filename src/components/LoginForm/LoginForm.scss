.login-form {
  min-width: 400px;

  input {
    background-color: #fff !important;
    &:focus {
      color: #999 !important;
    }
  }

}

.login-button {
    background-color: #9fd533 !important;
    border-color: #65873d !important;
    &:hover {
        background-color: #86b300 !important;
      }
}

.login-username {
  padding-left: 35px;
}

.login-password {
  padding-left: 35px;
}

.login-icon {
  position: relative;
  top: 30px;
  left: 10px;
}

.form-invalid {
  border-color: red;
}

.form-group {
  margin-top: -20px;
}