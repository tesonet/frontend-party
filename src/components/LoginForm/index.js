import React from "react";
import { connect } from "react-redux";
import { onLoginSubmit } from "../../thunks/userThunks/loginThunk";
import { useFormik } from "formik";
import Spinner from "../../assets/svg/spinner.svg";
import IconUsername from "../../assets/svg/user.svg";
import IconPassword from "../../assets/svg/lock.svg";

const LoginForm = ({ error, loading, history, onLogin }) => {
  const formik = useFormik({
    initialValues: {
      username: "",
      password: ""
    },
    onSubmit: values => {
      const { username, password } = values;
      onLogin({ username, password }, history);
    }
  });

  return (
    <form onSubmit={formik.handleSubmit} className="login-form">
      <div className="login-form__input-field">
        <img src={IconUsername} alt="" />
        <input
          id="username"
          name="username"
          type="text"
          onChange={formik.handleChange}
          value={formik.values.username}
          placeholder="Username"
          required
        />
      </div>
      <div className="login-form__input-field">
        <img src={IconPassword} alt="" />
        <input
          id="password"
          name="password"
          type="Password"
          onChange={formik.handleChange}
          values={formik.values.password}
          placeholder="Password"
          required
        />
      </div>
      <button
        type="submit"
        style={{ display: "flex", justifyContent: "center" }}
        disabled={loading}
      >
        {loading ? (
          <img
            src={Spinner}
            style={{ height: "100%", width: "16px" }}
            alt="Loading icon"
          />
        ) : (
          "Log In"
        )}
      </button>
      <p className="login-form__error">{error && error}</p>
    </form>
  );
};

const mapStateToProps = state => state.login;

const mapDispatchToProps = dispatch => ({
  onLogin: (user, history) => dispatch(onLoginSubmit(user, history))
});

export default connect(mapStateToProps, mapDispatchToProps)(LoginForm);
