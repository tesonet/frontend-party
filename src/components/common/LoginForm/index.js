import React, {useState, useRef, createRef} from 'react'
import {useDispatch, useSelector} from 'react-redux'
import {loginUser} from '../../../modules/auth/actions'
import {getIsLoggingIn, getErrorMessage} from '../../../modules/auth/selectors'
import {createValidator, required, minLength} from '../../../utils/validation'
import Input from '../../../ui/Input'
import Button from '../../../ui/Button'
import Alert from '../../../ui/Alert'

const config = [
  {
    type: 'text',
    name: 'username',
    label: 'Username',
    icon: 'user',
  },
  {
    type: 'password',
    name: 'password',
    label: 'Password',
    icon: 'lock',
  },
]

const validate = createValidator({
  username: [required, minLength(3)],
  password: required,
})

// TODO: reuse hooks logic

export default () => {
  const dispatch = useDispatch()
  const isLoggingIn = useSelector(getIsLoggingIn)
  const errorMessage = useSelector(getErrorMessage)

  const inputsRef = useRef(config.map(() => createRef()))

  const [inputs, setInputState] = useState({
    username: '',
    password: '',
  })

  const [errors, setErrors] = useState({})

  const handleInputChange = event => {
    const {target} = event
    const {type, name} = target
    const value = type === 'checkbox' ? target.checked : target.value

    setInputState(prevState => {
      return {
        ...prevState,
        [name]: value,
      }
    })
  }

  const handleFormSubmit = event => {
    const inputErrors = validate(inputs) || {}
    const invalidInputNames = Object.keys(inputErrors)

    if (invalidInputNames.length) {
      setErrors(inputErrors)
      // focus 1st invalid field
      try {
        const invalidInputName = invalidInputNames[0]

        inputsRef.current.forEach(ref => {
          if (ref.current.name === invalidInputName) {
            ref.current.focus()
          }
        })
      } catch (error) {
        console.error(error)
      }
    } else {
      setErrors({})
      dispatch(loginUser(inputs))
    }

    event.preventDefault()
  }

  return (
    <form onSubmit={handleFormSubmit}>
      {config.map((input, index) => {
        const {name} = input
        return (
          <Input
            key={name}
            ref={inputsRef.current[index]}
            type={input.type}
            name={input.name}
            label={input.label}
            value={inputs[name]}
            onChange={handleInputChange}
            startIcon={input.icon}
            disabled={isLoggingIn}
            error={!!errors[name]}
            helperText={errors[name]}
            mb={3}
          />
        )
      })}
      {!!errorMessage && (
        <Alert data-test-id="login-error" variant="danger" mb={3}>
          {errorMessage}
        </Alert>
      )}
      <Button
        variant="contained"
        size="medium"
        type="submit"
        loading={isLoggingIn}
        fullWidth>
        Log In
      </Button>
    </form>
  )
}
