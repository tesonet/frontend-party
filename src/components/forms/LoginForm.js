import React from 'react';
import {Form,Button,Input,Message} from 'semantic-ui-react';
import Validator from 'validator';
import PropTypes from 'prop-types';
import ErrorMessage from '../messages/ErrorMessage';




class LoginForm extends React.Component{

    state = {
        data : {
            email : '',
            password : ''
        },
        loading : false,
        errors : {}
    }

    onChange = e => this.setState({
        data : {...this.state.data, [e.target.name] : e.target.value }
    });

    onSubmit = () => {
        const errors = this.validate(this.state.data);
        this.setState({
            errors
        });
        if (Object.keys(errors).length === 0){
            this.setState({ loading : true});
            this.props
            .submit(this.state.data)
            .catch(err => this.setState({
                errors : err.response.data.errors, loading: false
            }));
            
        } 
    };

    validate = data => {
        const errors = {};
        if (!Validator.isEmail(data.email)) errors.email = "Invalid Email"
        if (!data.password) errors.password = "This can not be empty";
        return errors;
    }

    render(){

        let ButtonStyle = {
            width: "360px",
            height: "56px",
            borderRadius: "5px",
            backgroundColor: "#9fd533",
            color : "#ffff",
        }  
        
        const { data, errors, loading } = this.state;

        return(
            <Form onSubmit={this.onSubmit} loading= {loading}>   
                {errors.global && (
                    <Message negative>
                <Message.Header>Something went wrong :(</Message.Header>
                <p>{errors.global}</p>
                </Message>  
                

            )}
                <Form.Field error = {!!errors.email} >
                    <Input
                        icon='user' 
                        iconPosition='left'
                        type = "email"
                        id = "email"
                        name = "email"
                        value = {data.email}
                        placeholder = "Username"
                        onChange = {this.onChange}
                    />
                    {errors.email && <ErrorMessage text = { errors.email } />}
                </Form.Field>

                <Form.Field error = {!!errors.password}>
                    <Input
                        icon='lock' 
                        iconPosition='left'
                        type = "password"
                        id = "password"
                        name = "password"
                        value = {data.password}
                        placeholder = "Password"
                        onChange = {this.onChange}
                    />
                    {errors.password && <ErrorMessage text = { errors.password } />}
                </Form.Field>
                 <Button style = {ButtonStyle}>Login</Button>
            </Form>
        );
    }
}

LoginForm.propTypes = {
    submit : PropTypes.func.isRequired
}

export default LoginForm;