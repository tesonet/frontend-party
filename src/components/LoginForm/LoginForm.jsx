import React, { useState } from 'react';
import { PropTypes } from 'prop-types';
import userIcon from "../../assets/icons/ico-user.png";
import passwordIcon from "../../assets/icons/ico-lock.png";
import './LoginForm.scss';

const LoginForm = ({ submit, error }) => {
    const [username, setUserame] = useState("");
    const [password, setPassword] = useState("");

    const handleSubmit = event => {
        event.preventDefault();
        submit({ username, password });
    };

    return (
        <form className="login-form" onSubmit={e => handleSubmit(e)}>
            <div className="form-group">
                <img alt="username" className="login-icon" src={userIcon} />
                <input
                    type="text"
                    className={`form-control login-username ${error ? "form-invalid" : ""}`}
                    onChange={e => setUserame(e.target.value)}
                    value={username}
                    required
                    placeholder="Username" />
            </div>
            <div className="form-group">
                <img alt="password" className="login-icon" src={passwordIcon} />
                <input
                    type="password"
                    value={password}
                    onChange={e => setPassword(e.target.value)}
                    className={`form-control login-password ${error ? "form-invalid" : ""}`}
                    required
                    placeholder="Password" />
            </div>
            <button type="submit" className="btn btn-primary btn-block login-button" onSubmit={e => handleSubmit(e)}>Log In</button>
        </form>
    )
}

LoginForm.defaultProps = {
    error: false
};

LoginForm.propTypes = {
    submit: PropTypes.func.isRequired,
    error: PropTypes.bool
}

export default LoginForm;