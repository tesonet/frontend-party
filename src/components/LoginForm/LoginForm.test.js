
import * as React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16'
import LoginForm from "./LoginForm";

configure({ adapter: new Adapter() });

describe('Login Form', () => {

    it('should render component correctly', () => {
        const mockFn = jest.fn();
        const wrapper = shallow(<LoginForm submit={mockFn} />);
        expect(wrapper).toMatchSnapshot();
    });

    it('should render form', () => {
        const mockFn = jest.fn();
        const wrapper = shallow(<LoginForm submit={mockFn} />);
        expect(wrapper.find('.login-form').length).toBe(1);
    });

    it('should render username input', () => {
        const mockFn = jest.fn();
        const wrapper = shallow(<LoginForm submit={mockFn} />);
        expect(wrapper.find('.login-username').length).toBe(1);
    });

    it('should render password input', () => {
        const mockFn = jest.fn();
        const wrapper = shallow(<LoginForm submit={mockFn} />);
        expect(wrapper.find('.login-password').length).toBe(1);
    });

    it('should call submit callback', () => {
        const mockFn = jest.fn();
        const wrapper = shallow(<LoginForm submit={mockFn} />);
        wrapper.find('form').simulate('submit', { preventDefault: mockFn });
        expect(mockFn).toHaveBeenCalled();
    });

});