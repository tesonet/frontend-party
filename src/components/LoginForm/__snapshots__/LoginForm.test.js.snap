// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Form should render component correctly 1`] = `
<form
  className="login-form"
  onSubmit={[Function]}
>
  <div
    className="form-group"
  >
    <img
      alt="username"
      className="login-icon"
      src={null}
    />
    <input
      className="form-control login-username "
      onChange={[Function]}
      placeholder="Username"
      required={true}
      type="text"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <img
      alt="password"
      className="login-icon"
      src={null}
    />
    <input
      className="form-control login-password "
      onChange={[Function]}
      placeholder="Password"
      required={true}
      type="password"
      value=""
    />
  </div>
  <button
    className="btn btn-primary btn-block login-button"
    onSubmit={[Function]}
    type="submit"
  >
    Log In
  </button>
</form>
`;
